File: core/services/fsp_tracing/area/fsp_area_tracer_impl.cc
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "core/services/fsp_tracing/area/fsp_area_tracer_impl.h"

#include <algorithm>
#include <bitset>
#include <utility>

#include "core/services/fsp_tracing/base/fsp_snapshot.h"

namespace lynx {
namespace tasm {
namespace timing {

struct FSPAreaSnapshot : FSPSnapshot {
  static constexpr size_t X_PROJECTIONS_LEN = 256;
  static constexpr size_t Y_PROJECTIONS_LEN = 512;
  static constexpr size_t PROJECTIONS_LEN =
      X_PROJECTIONS_LEN * Y_PROJECTIONS_LEN;

  std::bitset<PROJECTIONS_LEN> projections;

  long last_ui_sign = -1;
  double last_change_timestamp;
  lynx::base::geometry::IntSize container_size;
  size_t projection_area;
};

bool DiffAreaSnapshots(FSPAreaSnapshot& current, FSPAreaSnapshot* previous,
                       const FSPAreaConfig& config);

std::unique_ptr<FSPSnapshot> FSPAreaTracer::CreateSnapshot() {
  return std::make_unique<FSPAreaSnapshot>();
}

FSPResult FSPAreaTracer::CaptureSnapshotImpl(
    FSPSnapshot& snapshot, lynx::base::geometry::IntSize container_size) {
  auto& ss = static_cast<FSPAreaSnapshot&>(snapshot);
  ss = {};

  if (!container_size.IsEmpty() && container_size.Width() <= UINT16_MAX &&
      container_size.Height() <= UINT16_MAX) {
    // Check container size to avoid overflow in the mulplications below.
    ss.container_size = container_size;
  }

  callback_(ss);
  bool is_stable = DiffAreaSnapshots(
      ss, static_cast<FSPAreaSnapshot*>(previous_snapshot_.get()), config_);
  return ss.GetResult(is_stable);
}

void FSPAreaTracer::FillSnapshotImpl(FSPSnapshot& snapshot,
                                     const FSPContentInfo& info) {
  auto& ss = static_cast<FSPAreaSnapshot&>(snapshot);
  if (info.rect.IsEmpty() || ss.container_size.IsEmpty()) {
    return;
  }

  // Only update last_ui_sign when we have a newer content timestamp
  if (info.change_timestamp >= ss.last_change_timestamp) {
    ss.last_ui_sign = info.ui_sign;
    ss.last_change_timestamp = info.change_timestamp;
  }

  auto w = ss.container_size.Width();
  auto h = ss.container_size.Height();

  auto min_x = static_cast<size_t>(std::clamp(info.rect.X(), 0, w));
  min_x = min_x * FSPAreaSnapshot::X_PROJECTIONS_LEN / w;
  min_x = std::min(min_x, FSPAreaSnapshot::X_PROJECTIONS_LEN - 1);

  auto max_x = static_cast<size_t>(std::clamp(info.rect.MaxX(), 0, w));
  max_x = max_x * FSPAreaSnapshot::X_PROJECTIONS_LEN / w;
  max_x = std::min(max_x, FSPAreaSnapshot::X_PROJECTIONS_LEN - 1);

  auto min_y = static_cast<size_t>(std::clamp(info.rect.Y(), 0, h));
  min_y = min_y * FSPAreaSnapshot::Y_PROJECTIONS_LEN / h;
  min_y = std::min(min_y, FSPAreaSnapshot::Y_PROJECTIONS_LEN - 1);

  auto max_y = static_cast<size_t>(std::clamp(info.rect.MaxY(), 0, h));
  max_y = max_y * FSPAreaSnapshot::Y_PROJECTIONS_LEN / h;
  max_y = std::min(max_y, FSPAreaSnapshot::Y_PROJECTIONS_LEN - 1);

  for (auto y = min_y; y <= max_y; y++) {
    for (auto x = min_x; x <= max_x; x++) {
      ss.projections[y * FSPAreaSnapshot::X_PROJECTIONS_LEN + x] = true;
    }
  }
}

bool DiffAreaSnapshots(FSPAreaSnapshot& current, FSPAreaSnapshot* previous,
                       const FSPAreaConfig& config) {
  auto container_a = static_cast<size_t>(current.container_size.Width()) *
                     static_cast<size_t>(current.container_size.Height());
  auto projection_a = current.projections.count() * container_a /
                      FSPAreaSnapshot::PROJECTIONS_LEN;
  current.projection_area = projection_a;

  // check against completion rates.
  auto min_a = static_cast<long>(static_cast<double>(container_a) *
                                 config.minimum_completion_rate);
  if ((min_a > 0) && (projection_a < static_cast<size_t>(min_a))) {
    return false;
  }

  if (!previous) {
    // Initial snapshot always considered stable.
    return true;
  }

  // diff area changes.(fast)
  auto diff_a = projection_a - previous->projection_area;
  if (diff_a >= config.acceptable_difference_per_snapshot) {
    return false;
  }

  // diff interval in miliseconds
  auto diff_t =
      (current.last_change_timestamp - previous->last_change_timestamp) *
      1000.0;

  // Check if the interval is less than the minimum interval.
  if ((config.minimum_diff_interval_ms > 0.0) &&
      (diff_t < config.minimum_diff_interval_ms)) {
    return false;
  }

  // diff area change rates.(fast)
  auto rate_a = static_cast<double>(diff_a) / diff_t;
  if (rate_a >= config.acceptable_difference_per_ms) {
    return false;
  }

  // diff content change rates.(slow)
  auto diff_p = (current.projections ^ previous->projections).count() *
                container_a / FSPAreaSnapshot::PROJECTIONS_LEN;
  auto rate_p = static_cast<double>(diff_p) / diff_t;
  bool is_stable = !(rate_p >= config.acceptable_difference_per_ms);
  return is_stable;
}

}  // namespace timing
}  // namespace tasm
}  // namespace lynx

File: platform/android/lynx_android/src/main/java/com/lynx/tasm/performance/fsptracing/area/FSPAreaTracerImpl.java
// Copyright 2025 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
package com.lynx.tasm.performance.fsptracing.area;

import android.graphics.Rect;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.annotation.RestrictTo;
import com.lynx.tasm.performance.fsptracing.FSPContentInfo;
import com.lynx.tasm.performance.fsptracing.FSPSnapshot;
import com.lynx.tasm.performance.fsptracing.IFSPTracerImpl;
import com.lynx.tasm.performance.fsptracing.area.FSPAreaConfig;
import com.lynx.tasm.performance.fsptracing.area.FSPAreaSnapshot;
import java.util.ArrayList;
import java.util.BitSet;

/**
 * Implementation of FSP tracer using area projection method.
 */
@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
@RestrictTo(RestrictTo.Scope.LIBRARY)
public class FSPAreaTracerImpl implements IFSPTracerImpl {
  private final FSPAreaConfig mConfig;

  public FSPAreaTracerImpl(FSPAreaConfig config) {
    mConfig = config;
  }

  @Override
  public FSPSnapshot createSnapshot() {
    return new FSPAreaSnapshot();
  }

  @Override
  public void fillSnapshot(@NonNull FSPSnapshot snapshot, @NonNull FSPContentInfo contentInfo) {
    assert snapshot instanceof FSPAreaSnapshot;

    Rect contentRect = contentInfo.getRect();
    int uiSign = contentInfo.getUiSign();
    long changeTimestampNs = contentInfo.getChangeTimestampNs();

    if (contentRect.isEmpty() || snapshot.getContainerWidth() <= 0
        || snapshot.getContainerHeight() <= 0) {
      return;
    }

    snapshot.updateTimestamp(changeTimestampNs, uiSign);

    int w = snapshot.getContainerWidth();
    int h = snapshot.getContainerHeight();

    FSPAreaSnapshot areaSnapshot = (FSPAreaSnapshot) snapshot;
    BitSet projections = areaSnapshot.getProjections();

    int minX = Math.max(0, Math.min(contentRect.left, w));
    minX = minX * FSPAreaSnapshot.X_PROJECTIONS_LEN / w;
    minX = Math.min(minX, FSPAreaSnapshot.X_PROJECTIONS_LEN - 1);

    int maxX = Math.max(0, Math.min(contentRect.right, w));
    maxX = maxX * FSPAreaSnapshot.X_PROJECTIONS_LEN / w;
    maxX = Math.min(maxX, FSPAreaSnapshot.X_PROJECTIONS_LEN - 1);

    int minY = Math.max(0, Math.min(contentRect.top, h));
    minY = minY * FSPAreaSnapshot.Y_PROJECTIONS_LEN / h;
    minY = Math.min(minY, FSPAreaSnapshot.Y_PROJECTIONS_LEN - 1);

    int maxY = Math.max(0, Math.min(contentRect.bottom, h));
    maxY = maxY * FSPAreaSnapshot.Y_PROJECTIONS_LEN / h;
    maxY = Math.min(maxY, FSPAreaSnapshot.Y_PROJECTIONS_LEN - 1);
    
    for (int y = minY; y <= maxY; y++) {
      for (int x = minX; x <= maxX; x++) {
        projections.set(y * FSPAreaSnapshot.X_PROJECTIONS_LEN + x);
      }
    }
  }

  @Override
  public boolean diffSnapshots(@NonNull FSPSnapshot current, FSPSnapshot previous) {
    assert current instanceof FSPAreaSnapshot;
    assert previous == null || previous instanceof FSPAreaSnapshot;

    FSPAreaSnapshot currentArea = (FSPAreaSnapshot) current;

    long containerArea = (long) currentArea.getContainerWidth() * currentArea.getContainerHeight();
    int projectionArea = (int) (currentArea.getProjections().cardinality() * containerArea
        / FSPAreaSnapshot.PROJECTIONS_LEN);
    currentArea.setProjectionArea(projectionArea);

    // Check against completion rates
    double compRate = (double) projectionArea / containerArea;
    if (compRate < mConfig.getMinimumCompletionRate()) {
      return false;
    }

    if (previous == null) {
      // Initial snapshot always considered stable.
      return true;
    }

    FSPAreaSnapshot prevArea = (FSPAreaSnapshot) previous;

    // diff area changes
    int diffArea = projectionArea - prevArea.getProjectionArea();
    if (diffArea >= mConfig.getAcceptableDifferencePerSnapshot()) {
      return false;
    }

    double diffT =
        (currentArea.getLastChangeTimestampNs() - prevArea.getLastChangeTimestampNs()) / 1000_000.0;

    // Check if the interval is less than the minimum interval.
    if (diffT <= mConfig.getMinimumDiffIntervalMs()) {
      return false;
    }

    // diff area change rates
    double rateArea = diffArea / diffT;
    if (rateArea >= mConfig.getAcceptableDifferencePerMs()) {
      return false;
    }

    // diff content change rates
    BitSet diffProjections = (BitSet) currentArea.getProjections().clone();
    diffProjections.xor(prevArea.getProjections());
    int diffContent =
        (int) (diffProjections.cardinality() * containerArea / FSPAreaSnapshot.PROJECTIONS_LEN);
    double rateContent = diffContent / diffT;

    if (rateContent >= mConfig.getAcceptableDifferencePerMs()) {
      return false;
    }

    return true;
  }

  @Override
  public ArrayList<String> inspectSnapshot(@NonNull FSPSnapshot current, FSPSnapshot previous) {
    assert current instanceof FSPAreaSnapshot;
    assert previous == null || previous instanceof FSPAreaSnapshot;

    // TODO: Implement inspection of snapshot detail.
    return new ArrayList<>();
  }
}
